File: src/ormai/core/dsl.py
"""
Query DSL schemas for OrmAI.

These Pydantic models define the structured query language that agents use
to interact with the database. The DSL is designed to be safe, expressive,
and easily validated.
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class FilterOp(str, Enum):
    """Supported filter operators."""

    EQ = "eq"
    NE = "ne"
    LT = "lt"
    LTE = "lte"
    GT = "gt"
    GTE = "gte"
    IN = "in"
    NOT_IN = "not_in"
    IS_NULL = "is_null"
    CONTAINS = "contains"
    STARTSWITH = "startswith"
    ENDSWITH = "endswith"
    BETWEEN = "between"


class OrderDirection(str, Enum):
    """Sort order direction."""

    ASC = "asc"
    DESC = "desc"


class FilterClause(BaseModel):
    """
    A single filter condition.

    Examples:
        {"field": "status", "op": "eq", "value": "active"}
        {"field": "created_at", "op": "gte", "value": "2024-01-01"}
        {"field": "id", "op": "in", "value": [1, 2, 3]}
    """

    field: str = Field(..., description="The field name to filter on")
    op: FilterOp = Field(..., description="The filter operator")
    value: Any = Field(..., description="The value to compare against")

    model_config = {"frozen": True}

    @field_validator("field")
    @classmethod
    def validate_field_name(cls, v: str) -> str:
        """Ensure field name is not empty and doesn't contain SQL injection attempts."""
        if not v or not v.strip():
            raise ValueError("Field name cannot be empty")
        # Basic protection against SQL injection in field names
        if any(char in v for char in [";", "--", "/*", "*/", "'", '"']):
            raise ValueError("Invalid characters in field name")
        return v.strip()


class OrderClause(BaseModel):
    """
    A single order/sort clause.

    Example:
        {"field": "created_at", "direction": "desc"}
    """

    field: str = Field(..., description="The field name to order by")
    direction: OrderDirection = Field(
        default=OrderDirection.ASC, description="Sort direction"
    )

    model_config = {"frozen": True}


class IncludeClause(BaseModel):
    """
    Specifies a relation to include in the query results.

    Example:
        {"relation": "orders", "select": ["id", "total", "status"]}
    """

    relation: str = Field(..., description="The relation name to include")
    select: list[str] | None = Field(
        default=None, description="Fields to select from the relation"
    )
    where: list[FilterClause] | None = Field(
        default=None, description="Filters to apply to the relation"
    )
    take: int | None = Field(
        default=None, ge=1, description="Max items to include from relation (limited by budget)"
    )

    model_config = {"frozen": True}


class QueryRequest(BaseModel):
    """
    A structured query request.

    This is the primary DSL for querying data through OrmAI. It supports:
    - Field selection
    - Filtering with various operators
    - Ordering
    - Cursor-based pagination
    - Relation includes

    Example:
        {
            "model": "Order",
            "select": ["id", "total", "status", "created_at"],
            "where": [
                {"field": "status", "op": "eq", "value": "pending"},
                {"field": "created_at", "op": "gte", "value": "2024-01-01"}
            ],
            "order_by": [{"field": "created_at", "direction": "desc"}],
            "take": 25,
            "include": [{"relation": "customer", "select": ["id", "name"]}]
        }
    """

    model: str = Field(..., description="The model/table name to query")
    select: list[str] | None = Field(
        default=None, description="Fields to select (None means all allowed fields)"
    )
    where: list[FilterClause] | None = Field(
        default=None, description="Filter conditions (AND-ed together)"
    )
    order_by: list[OrderClause] | None = Field(
        default=None, description="Sort order"
    )
    take: int = Field(
        default=25, ge=1, description="Maximum number of rows to return (limited by budget)"
    )
    cursor: str | None = Field(
        default=None, description="Pagination cursor from previous response"
    )
    include: list[IncludeClause] | None = Field(
        default=None, description="Relations to include"
    )

    model_config = {"frozen": True}


class GetRequest(BaseModel):
    """
    A request to get a single record by primary key.

    Example:
        {
            "model": "Order",
            "id": 123,
            "select": ["id", "total", "status"],
            "include": [{"relation": "items"}]
        }
    """

    model: str = Field(..., description="The model/table name")
    id: Any = Field(..., description="The primary key value")
    select: list[str] | None = Field(
        default=None, description="Fields to select"
    )
    include: list[IncludeClause] | None = Field(
        default=None, description="Relations to include"
    )

    model_config = {"frozen": True}


class AggregateRequest(BaseModel):
    """
    A request to perform an aggregation.

    Example:
        {
            "model": "Order",
            "operation": "sum",
            "field": "total",
            "where": [{"field": "status", "op": "eq", "value": "completed"}]
        }
    """

    model: str = Field(..., description="The model/table name")
    operation: str = Field(..., description="Aggregation operation: count, sum, avg, min, max")
    field: str | None = Field(
        default=None, description="Field to aggregate (required for sum/avg/min/max)"
    )
    where: list[FilterClause] | None = Field(
        default=None, description="Filter conditions"
    )

    model_config = {"frozen": True}

    @field_validator("operation")
    @classmethod
    def validate_operation(cls, v: str) -> str:
        """Validate aggregation operation."""
        allowed = {"count", "sum", "avg", "min", "max"}
        if v.lower() not in allowed:
            raise ValueError(f"Operation must be one of: {', '.join(allowed)}")
        return v.lower()


class QueryResult(BaseModel):
    """
    Result of a query operation.
    """

    data: list[dict[str, Any]] = Field(default_factory=list)
    next_cursor: str | None = Field(default=None)
    has_more: bool = Field(default=False)
    total_count: int | None = Field(default=None)

    model_config = {"frozen": True}


class GetResult(BaseModel):
    """
    Result of a get operation.
    """

    data: dict[str, Any] | None = Field(default=None)
    found: bool = Field(default=False)

    model_config = {"frozen": True}


class AggregateResult(BaseModel):
    """
    Result of an aggregation operation.
    """

    value: Any = Field(default=None)
    operation: str
    field: str | None = None
    row_count: int = Field(default=0)

    model_config = {"frozen": True}


# =============================================================================
# MUTATION DSL (Phase 2)
# =============================================================================


class CreateRequest(BaseModel):
    """
    A request to create a new record.

    Example:
        {
            "model": "Order",
            "data": {
                "customer_id": 123,
                "total": 99.99,
                "status": "pending"
            },
            "reason": "Customer placed order via checkout flow"
        }
    """

    model: str = Field(..., description="The model/table name")
    data: dict[str, Any] = Field(..., description="Field values for the new record")
    reason: str | None = Field(
        default=None,
        description="Reason for the mutation (required by some policies)",
    )
    return_fields: list[str] | None = Field(
        default=None,
        description="Fields to return after creation (None means all allowed)",
    )

    model_config = {"frozen": True}


class UpdateRequest(BaseModel):
    """
    A request to update a record by primary key.

    Example:
        {
            "model": "Order",
            "id": 123,
            "data": {"status": "shipped"},
            "reason": "Order shipped by warehouse"
        }
    """

    model: str = Field(..., description="The model/table name")
    id: Any = Field(..., description="The primary key of the record to update")
    data: dict[str, Any] = Field(..., description="Fields to update")
    reason: str | None = Field(
        default=None,
        description="Reason for the mutation",
    )
    return_fields: list[str] | None = Field(
        default=None,
        description="Fields to return after update",
    )

    model_config = {"frozen": True}


class DeleteRequest(BaseModel):
    """
    A request to delete a record by primary key.

    By default, this performs a soft delete if the model has a soft delete field.
    Hard deletes require explicit policy permission.

    Example:
        {
            "model": "Order",
            "id": 123,
            "reason": "Customer requested cancellation",
            "hard": false
        }
    """

    model: str = Field(..., description="The model/table name")
    id: Any = Field(..., description="The primary key of the record to delete")
    reason: str | None = Field(
        default=None,
        description="Reason for the deletion",
    )
    hard: bool = Field(
        default=False,
        description="If True, perform hard delete instead of soft delete",
    )

    model_config = {"frozen": True}


class BulkUpdateRequest(BaseModel):
    """
    A request to update multiple records by their primary keys.

    This is safer than a filter-based bulk update because it requires
    explicit identification of each record to update.

    Example:
        {
            "model": "Order",
            "ids": [123, 124, 125],
            "data": {"status": "cancelled"},
            "reason": "Batch cancellation due to supplier issue"
        }
    """

    model: str = Field(..., description="The model/table name")
    ids: list[Any] = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Primary keys of records to update",
    )
    data: dict[str, Any] = Field(..., description="Fields to update on all records")
    reason: str | None = Field(
        default=None,
        description="Reason for the mutation",
    )

    model_config = {"frozen": True}


class CreateResult(BaseModel):
    """
    Result of a create operation.
    """

    data: dict[str, Any] = Field(default_factory=dict)
    id: Any = Field(default=None, description="Primary key of created record")
    success: bool = Field(default=True)

    model_config = {"frozen": True}


class UpdateResult(BaseModel):
    """
    Result of an update operation.
    """

    data: dict[str, Any] | None = Field(default=None)
    success: bool = Field(default=True)
    found: bool = Field(default=True)

    model_config = {"frozen": True}


class DeleteResult(BaseModel):
    """
    Result of a delete operation.
    """

    success: bool = Field(default=True)
    found: bool = Field(default=True)
    soft_deleted: bool = Field(default=True, description="True if soft delete was used")

    model_config = {"frozen": True}


class BulkUpdateResult(BaseModel):
    """
    Result of a bulk update operation.
    """

    updated_count: int = Field(default=0)
    success: bool = Field(default=True)
    failed_ids: list[Any] = Field(
        default_factory=list,
        description="IDs that failed to update (if any)",
    )

    model_config = {"frozen": True}

File: ormai-ts/src/core/dsl.ts
/**
 * Query DSL schemas for OrmAI.
 *
 * These Zod schemas define the structured query language that agents use
 * to interact with the database. The DSL is designed to be safe, expressive,
 * and easily validated.
 */

import { z } from 'zod';

// =============================================================================
// FILTER OPERATORS
// =============================================================================

/**
 * Supported filter operators.
 */
export const FilterOpSchema = z.enum([
  'eq',
  'ne',
  'lt',
  'lte',
  'gt',
  'gte',
  'in',
  'not_in',
  'is_null',
  'contains',
  'startswith',
  'endswith',
  'between',
]);

export type FilterOp = z.infer<typeof FilterOpSchema>;

/**
 * Sort order direction.
 */
export const OrderDirectionSchema = z.enum(['asc', 'desc']);

export type OrderDirection = z.infer<typeof OrderDirectionSchema>;

// =============================================================================
// CLAUSE SCHEMAS
// =============================================================================

/**
 * Validate field name to prevent injection.
 */
const fieldNameSchema = z
  .string()
  .min(1, 'Field name cannot be empty')
  .refine(
    (v) => !v.includes(';') && !v.includes('--') && !v.includes('/*') && !v.includes('*/'),
    { message: 'Invalid characters in field name' }
  )
  .transform((v) => v.trim());

/**
 * A single filter condition.
 *
 * @example
 * { field: "status", op: "eq", value: "active" }
 * { field: "created_at", op: "gte", value: "2024-01-01" }
 * { field: "id", op: "in", value: [1, 2, 3] }
 */
export const FilterClauseSchema = z
  .object({
    field: fieldNameSchema.describe('The field name to filter on'),
    op: FilterOpSchema.describe('The filter operator'),
    value: z.unknown().describe('The value to compare against'),
  })
  .readonly();

export type FilterClause = z.infer<typeof FilterClauseSchema>;

/**
 * A single order/sort clause.
 *
 * @example
 * { field: "created_at", direction: "desc" }
 */
export const OrderClauseSchema = z
  .object({
    field: fieldNameSchema.describe('The field name to order by'),
    direction: OrderDirectionSchema.default('asc').describe('Sort direction'),
  })
  .readonly();

export type OrderClause = z.infer<typeof OrderClauseSchema>;

/**
 * Specifies a relation to include in the query results.
 *
 * @example
 * { relation: "orders", select: ["id", "total", "status"] }
 */
export const IncludeClauseSchema = z
  .object({
    relation: z.string().min(1).describe('The relation name to include'),
    select: z.array(z.string()).optional().describe('Fields to select from the relation'),
    where: z.array(z.lazy(() => FilterClauseSchema)).optional().describe('Filters to apply to the relation'),
    take: z.number().int().min(1).max(100).optional().describe('Max items to include from relation'),
  })
  .readonly();

export type IncludeClause = z.infer<typeof IncludeClauseSchema>;

// =============================================================================
// QUERY REQUEST/RESULT SCHEMAS
// =============================================================================

/**
 * A structured query request.
 *
 * This is the primary DSL for querying data through OrmAI. It supports:
 * - Field selection
 * - Filtering with various operators
 * - Ordering
 * - Cursor-based pagination
 * - Relation includes
 *
 * @example
 * {
 *   model: "Order",
 *   select: ["id", "total", "status", "created_at"],
 *   where: [
 *     { field: "status", op: "eq", value: "pending" },
 *     { field: "created_at", op: "gte", value: "2024-01-01" }
 *   ],
 *   orderBy: [{ field: "created_at", direction: "desc" }],
 *   take: 25,
 *   include: [{ relation: "customer", select: ["id", "name"] }]
 * }
 */
export const QueryRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name to query'),
    select: z.array(z.string()).optional().describe('Fields to select (undefined means all allowed fields)'),
    where: z.array(FilterClauseSchema).optional().describe('Filter conditions (AND-ed together)'),
    orderBy: z.array(OrderClauseSchema).optional().describe('Sort order'),
    take: z.number().int().min(1).max(100).default(25).describe('Maximum number of rows to return'),
    cursor: z.string().optional().describe('Pagination cursor from previous response'),
    include: z.array(IncludeClauseSchema).optional().describe('Relations to include'),
  })
  .readonly();

export type QueryRequest = z.infer<typeof QueryRequestSchema>;

/**
 * A request to get a single record by primary key.
 *
 * @example
 * {
 *   model: "Order",
 *   id: 123,
 *   select: ["id", "total", "status"],
 *   include: [{ relation: "items" }]
 * }
 */
export const GetRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name'),
    id: z.unknown().describe('The primary key value'),
    select: z.array(z.string()).optional().describe('Fields to select'),
    include: z.array(IncludeClauseSchema).optional().describe('Relations to include'),
  })
  .readonly();

export type GetRequest = z.infer<typeof GetRequestSchema>;

/**
 * A request to perform an aggregation.
 *
 * @example
 * {
 *   model: "Order",
 *   operation: "sum",
 *   field: "total",
 *   where: [{ field: "status", op: "eq", value: "completed" }]
 * }
 */
export const AggregateRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name'),
    operation: z
      .enum(['count', 'sum', 'avg', 'min', 'max'])
      .describe('Aggregation operation'),
    field: z.string().optional().describe('Field to aggregate (required for sum/avg/min/max)'),
    where: z.array(FilterClauseSchema).optional().describe('Filter conditions'),
  })
  .readonly();

export type AggregateRequest = z.infer<typeof AggregateRequestSchema>;

/**
 * Result of a query operation.
 */
export const QueryResultSchema = z
  .object({
    data: z.array(z.record(z.unknown())).default([]),
    nextCursor: z.string().nullable().default(null),
    hasMore: z.boolean().default(false),
    totalCount: z.number().int().nullable().default(null),
  })
  .readonly();

export type QueryResult = z.infer<typeof QueryResultSchema>;

/**
 * Result of a get operation.
 */
export const GetResultSchema = z
  .object({
    data: z.record(z.unknown()).nullable().default(null),
    found: z.boolean().default(false),
  })
  .readonly();

export type GetResult = z.infer<typeof GetResultSchema>;

/**
 * Result of an aggregation operation.
 */
export const AggregateResultSchema = z
  .object({
    value: z.unknown().default(null),
    operation: z.string(),
    field: z.string().nullable().default(null),
    rowCount: z.number().int().default(0),
  })
  .readonly();

export type AggregateResult = z.infer<typeof AggregateResultSchema>;

// =============================================================================
// MUTATION REQUEST/RESULT SCHEMAS
// =============================================================================

/**
 * A request to create a new record.
 *
 * @example
 * {
 *   model: "Order",
 *   data: {
 *     customerId: 123,
 *     total: 99.99,
 *     status: "pending"
 *   },
 *   reason: "Customer placed order via checkout flow"
 * }
 */
export const CreateRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name'),
    data: z.record(z.unknown()).describe('Field values for the new record'),
    reason: z.string().optional().describe('Reason for the mutation (required by some policies)'),
    returnFields: z.array(z.string()).optional().describe('Fields to return after creation'),
  })
  .readonly();

export type CreateRequest = z.infer<typeof CreateRequestSchema>;

/**
 * A request to update a record by primary key.
 *
 * @example
 * {
 *   model: "Order",
 *   id: 123,
 *   data: { status: "shipped" },
 *   reason: "Order shipped by warehouse"
 * }
 */
export const UpdateRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name'),
    id: z.unknown().describe('The primary key of the record to update'),
    data: z.record(z.unknown()).describe('Fields to update'),
    reason: z.string().optional().describe('Reason for the mutation'),
    returnFields: z.array(z.string()).optional().describe('Fields to return after update'),
  })
  .readonly();

export type UpdateRequest = z.infer<typeof UpdateRequestSchema>;

/**
 * A request to delete a record by primary key.
 *
 * By default, this performs a soft delete if the model has a soft delete field.
 * Hard deletes require explicit policy permission.
 *
 * @example
 * {
 *   model: "Order",
 *   id: 123,
 *   reason: "Customer requested cancellation",
 *   hard: false
 * }
 */
export const DeleteRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name'),
    id: z.unknown().describe('The primary key of the record to delete'),
    reason: z.string().optional().describe('Reason for the deletion'),
    hard: z.boolean().default(false).describe('If true, perform hard delete instead of soft delete'),
  })
  .readonly();

export type DeleteRequest = z.infer<typeof DeleteRequestSchema>;

/**
 * A request to update multiple records by their primary keys.
 *
 * This is safer than a filter-based bulk update because it requires
 * explicit identification of each record to update.
 *
 * @example
 * {
 *   model: "Order",
 *   ids: [123, 124, 125],
 *   data: { status: "cancelled" },
 *   reason: "Batch cancellation due to supplier issue"
 * }
 */
export const BulkUpdateRequestSchema = z
  .object({
    model: z.string().min(1).describe('The model/table name'),
    ids: z
      .array(z.unknown())
      .min(1)
      .max(100)
      .describe('Primary keys of records to update'),
    data: z.record(z.unknown()).describe('Fields to update on all records'),
    reason: z.string().optional().describe('Reason for the mutation'),
  })
  .readonly();

export type BulkUpdateRequest = z.infer<typeof BulkUpdateRequestSchema>;

/**
 * Result of a create operation.
 */
export const CreateResultSchema = z
  .object({
    data: z.record(z.unknown()).default({}),
    id: z.unknown().describe('Primary key of created record'),
    success: z.boolean().default(true),
  })
  .readonly();

export type CreateResult = z.infer<typeof CreateResultSchema>;

/**
 * Result of an update operation.
 */
export const UpdateResultSchema = z
  .object({
    data: z.record(z.unknown()).nullable().default(null),
    success: z.boolean().default(true),
    found: z.boolean().default(true),
  })
  .readonly();

export type UpdateResult = z.infer<typeof UpdateResultSchema>;

/**
 * Result of a delete operation.
 */
export const DeleteResultSchema = z
  .object({
    success: z.boolean().default(true),
    found: z.boolean().default(true),
    softDeleted: z.boolean().default(true).describe('True if soft delete was used'),
  })
  .readonly();

export type DeleteResult = z.infer<typeof DeleteResultSchema>;

/**
 * Result of a bulk update operation.
 */
export const BulkUpdateResultSchema = z
  .object({
    updatedCount: z.number().int().default(0),
    success: z.boolean().default(true),
    failedIds: z.array(z.unknown()).default([]).describe('IDs that failed to update'),
  })
  .readonly();

export type BulkUpdateResult = z.infer<typeof BulkUpdateResultSchema>;

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

/**
 * Create a filter clause with type safety.
 */
export function filter(field: string, op: FilterOp, value: unknown): FilterClause {
  return FilterClauseSchema.parse({ field, op, value });
}

/**
 * Create an order clause with type safety.
 */
export function orderBy(field: string, direction: OrderDirection = 'asc'): OrderClause {
  return OrderClauseSchema.parse({ field, direction });
}

/**
 * Create an include clause with type safety.
 */
export function include(
  relation: string,
  opts?: { select?: string[]; where?: FilterClause[]; take?: number }
): IncludeClause {
  return IncludeClauseSchema.parse({ relation, ...opts });
}

/**
 * Union type for all request types.
 */
export type AnyRequest =
  | QueryRequest
  | GetRequest
  | AggregateRequest
  | CreateRequest
  | UpdateRequest
  | DeleteRequest
  | BulkUpdateRequest;

/**
 * Union type for all result types.
 */
export type AnyResult =
  | QueryResult
  | GetResult
  | AggregateResult
  | CreateResult
  | UpdateResult
  | DeleteResult
  | BulkUpdateResult;
